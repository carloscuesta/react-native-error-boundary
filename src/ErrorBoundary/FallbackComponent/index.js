// @flow
import React from 'react'
import {
  SafeAreaView,
  Text,
  TouchableOpacity,
  View
} from 'react-native'

import styles from './styles'

type Props = { error: Error, resetError: Function }

const FallbackComponent = (props: Props) => (
  <SafeAreaView style={styles.container}>
    <View style={styles.content}>
      <Text style={styles.title}>{'Oops!'}</Text>
      <Text style={styles.subtitle}>{'There\'s an error'}</Text>
      <Text style={styles.error}>{props.error.toString()}</Text>
      <TouchableOpacity style={styles.button} onPress={props.resetError}>
        <Text style={styles.buttonText}>{'Try again'}</Text>
      </TouchableOpacity>
    </View>
  </SafeAreaView>
)

export default FallbackComponent
