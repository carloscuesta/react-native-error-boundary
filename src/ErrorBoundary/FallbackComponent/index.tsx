import React from 'react'
import { Text, TouchableOpacity, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

import styles from './styles'

export type Props = { error: Error; resetError: () => void }

const FallbackComponent = (props: Props) => (
  <SafeAreaView style={styles.container}>
    <View style={styles.content}>
      <Text style={styles.title}>Oops!</Text>
      <Text style={styles.subtitle}>{"There's an error"}</Text>
      <Text style={styles.error}>{props.error.toString()}</Text>
      <TouchableOpacity style={styles.button} onPress={props.resetError}>
        <Text style={styles.buttonText}>Try again</Text>
      </TouchableOpacity>
    </View>
  </SafeAreaView>
)

export default FallbackComponent
