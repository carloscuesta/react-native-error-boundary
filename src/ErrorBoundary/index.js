// @flow
import React, { type Node, type ComponentType } from 'react'

import FallbackComponent from './FallbackComponent'

type Props = {
  children: Node,
  FallbackComponent: ComponentType<any>,
  onError?: Function
}

type State = { error: Error | null, hasError: boolean }

class ErrorBoundary extends React.Component<Props, State> {
  state = { error: null, hasError: false }

  static defaultProps = {
    FallbackComponent: FallbackComponent
  }

  static getDerivedStateFromError (error: Error) {
    return { error, hasError: true }
  }

  componentDidCatch (error: Error, info: { componentStack: string }) {
    if (typeof this.props.onError === 'function') {
      this.props.onError.call(this, error, info.componentStack)
    }
  }

  resetError: Function = () => {
    this.setState({ error: null, hasError: false })
  }

  render () {
    const { FallbackComponent } = this.props

    return this.state.hasError
      ? <FallbackComponent
        error={this.state.error}
        resetError={this.resetError}
      />
      : this.props.children
  }
}

export default ErrorBoundary
