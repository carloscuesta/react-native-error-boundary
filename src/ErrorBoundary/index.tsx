import React, { type ComponentType, type ReactNode } from 'react'

import FallbackComponent, {
  type Props as FallbackComponentProps
} from './FallbackComponent'

export type Props = {
  children: Exclude<NonNullable<ReactNode>, string | number | boolean>,
  FallbackComponent: ComponentType<FallbackComponentProps>,
  onError?: (error: Error, stackTrace: string) => void
}

type State = { error: Error | null }

class ErrorBoundary extends React.Component<Props, State> {
  state: State = { error: null }

  static defaultProps: { FallbackComponent: ComponentType<FallbackComponentProps> } = {
    FallbackComponent: FallbackComponent
  }

  static getDerivedStateFromError (error: Error): State {
    return { error }
  }

  componentDidCatch (error: Error, info: { componentStack: string }) {
    if (typeof this.props.onError === 'function') {
      this.props.onError(error, info.componentStack)
    }
  }

  resetError: () => void = () => {
    this.setState({ error: null })
  }

  render () {
    const { FallbackComponent } = this.props

    return this.state.error
      ? <FallbackComponent
          error={this.state.error}
          resetError={this.resetError}
        />
      : this.props.children
  }
}

export default ErrorBoundary
