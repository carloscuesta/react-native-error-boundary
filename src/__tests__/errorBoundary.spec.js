import React from 'react'
import { Text, View } from 'react-native'
import renderer from 'react-test-renderer'

import moduleIndex from '../index'
import ErrorBoundary from '../ErrorBoundary'

describe('ErrorBoundary', () => {
  let consoleErrorSpy
  let errorMock

  beforeAll(() => {
    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {})
    errorMock = new Error('This is a test error!')
  })

  afterAll(() => {
    consoleErrorSpy.mockRestore()
  })

  describe('Module index', () => {
    it('should default export ErrorBoundary', () => {
      expect(moduleIndex).toBeInstanceOf(Function)
    })
  })

  it('should render children when there are no errors', () => {
    const wrapper = renderer.create(
      <ErrorBoundary>
        <Text>Hey!</Text>
      </ErrorBoundary>
    )
    expect(wrapper.getInstance().state.error).toBe(null)
    expect(wrapper.getInstance().state.hasError).toBe(false)
    expect(wrapper).toMatchSnapshot()
  })

  it('should catch errors set the state and render the default FallbackComponent', () => {
    const wrapper = renderer.create(
      <ErrorBoundary>
        <View>{errorMock}</View>
      </ErrorBoundary>
    )
    expect(wrapper.getInstance().state.hasError).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('should catch errors set the state render the props.FallbackComponent and call props.onError', () => {
    const FallbackComponent = () => <Text>Error!</Text>
    const onError = jest.fn()
    const wrapper = renderer.create(
      <ErrorBoundary FallbackComponent={FallbackComponent} onError={onError}>
        <View>{errorMock}</View>
      </ErrorBoundary>
    )
    expect(wrapper.getInstance().state.error).not.toBeNull()
    expect(wrapper.getInstance().state.hasError).toBe(true)
    expect(onError).toHaveBeenCalledWith(expect.any(Error), expect.any(String))
    expect(wrapper).toMatchSnapshot()
  })

  it('should clear state when calling resetError', () => {
    const wrapper = renderer.create(
      <ErrorBoundary>
        <Text>No errors!</Text>
      </ErrorBoundary>
    )
    wrapper.getInstance().setState({ error: errorMock, hasError: true })
    wrapper.getInstance().resetError()
    expect(wrapper.getInstance().state.error).toBeNull()
    expect(wrapper.getInstance().state.hasError).toBe(false)
  })
})
