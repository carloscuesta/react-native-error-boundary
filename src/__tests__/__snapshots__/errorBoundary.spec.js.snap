// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should catch errors set the state and render the default FallbackComponent 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#fafafa",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "marginHorizontal": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 48,
          "fontWeight": "300",
          "paddingBottom": 16,
        }
      }
    >
      Oops!
    </Text>
    <Text
      style={
        Object {
          "fontSize": 32,
          "fontWeight": "800",
        }
      }
    >
      There's an error
    </Text>
    <Text
      style={
        Object {
          "paddingVertical": 16,
        }
      }
    >
      Invariant Violation: Objects are not valid as a React child (found: Error: This is a test error!). If you meant to render a collection of children, use an array instead.
    in View (created by View)
    in View (at errorBoundary.spec.js:41)
    in ErrorBoundary (at errorBoundary.spec.js:40)
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#2196f3",
          "borderRadius": 50,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "600",
            "textAlign": "center",
          }
        }
      >
        Try again
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`ErrorBoundary should catch errors set the state render the props.FallbackComponent and call props.onError 1`] = `
<Text>
  Error!
</Text>
`;

exports[`ErrorBoundary should render children when there are no errors 1`] = `
<Text>
  Hey!
</Text>
`;
